File: algorithms/java/src/test/java/leetcode/LC382LinkedListRandomNode.java
package leetcode.lc382_linked_list_random_node;

import org.junit.jupiter.api.Test;


import java.util.Random;

import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * https://leetcode.com/problems/linked-list-random-node/
 * #medium
 */
public final class LC382LinkedListRandomNode {
    public final class ListNode {
        int val;
        ListNode next;

        ListNode(final int x) {
            val = x;
        }
    }

    public final class Solution {
        private final ListNode head;
        private final Random random;

        public Solution(final ListNode head) {
            this.head = head;
            random = new Random();
        }

        public int getRandom() {
            ListNode result = null, current = head;
            for (int i = 1; current != null; i++, current = current.next) {
                if (random.nextInt(i) == 0) {
                    result = current;
                }
            }
            return result.val;
        }
    }

    @Test
    public void test_example() throws Exception {
        ListNode head = new ListNode(1);
        head.next = new ListNode(2);
        head.next.next = new ListNode(3);
        int[] counts = new int[3];
        Solution solution = new Solution(head);
        for (int i = 0; i < 10000; i++) {
            int value = solution.getRandom();
            assertTrue(1 <= value && value <= 3);
            counts[value - 1]++;
        }
        assertEquals(3, counts[0] / 1000);
        assertEquals(3, counts[1] / 1000);
        assertEquals(3, counts[2] / 1000);
        // Should use Chi-squared test
    }
}

File: algorithms/code/leetcode/lc382_linked_list_random_node/lc382_linked_list_random_node.py
# https://leetcode.com/problems/linked-list-random-node/

from __future__ import annotations

import random
import unittest


class ListNode:
    def __init__(
        self, val: int = 0, following: 'ListNode | None' = None
    ) -> None:
        self.val = val
        self.next = following


class Solution:
    def __init__(self, head: ListNode | None) -> None:
        self.__head = head

    def getRandom(self) -> int:
        result = None
        current = self.__head
        i = 1
        while current is not None:
            if random.SystemRandom().randint(0, i - 1) == 0:
                result = current
            i += 1
            current = current.next
        if result is not None:
            return result.val
        raise ValueError('Empty list')


class TestCode(unittest.TestCase):
    def test_example(self) -> None:
        head = ListNode(1)
        head.next = ListNode(2)
        head.next.next = ListNode(3)
        counts = [0] * 3
        solution = Solution(head)
        for _ in range(0, 10000):
            value = solution.getRandom()
            assert 1 <= value <= 3
            counts[value - 1] += 1
        assert counts[0] // 1000 == 3
        assert counts[1] // 1000 == 3
        assert counts[2] // 1000 == 3
        # Should use Chi-squared test

    def test_nothing(self) -> None:
        solution = Solution(None)
        try:
            solution.getRandom()
        except ValueError:
            pass
